### ISSUE TITLE
Interplay between custom validators and optional fields

### ISSUE THREAD

**Issue by baleksan** [asaskevich/govalidator#149]
As far as I understand and can surmise from the code,  **optional,required** annotations are ignored when custom validators are defined on the same field.

For instance, if I have defined a custom validator called version which allows for regex XX.XX.XXX on this field

``` go
    Version  string `json:"version" valid:"optional,version"`
```

I expect when I am validating my struct as 

``` go
    _, err = govalidator.ValidateStruct(<struct>)
```

empty string will be allowed, but it is not!

I wonder why the design decision was made this way? 

It is most inconvenient and requires users to always handle the empty field case explicitly. It is confusing that the validator which is a **part of a standard library**

``` go
    Email  string `json:"email" valid:"optional,email"`
```

will work correctly with empty string, but a **custom** validator 

``` go
    Version  string `json:"version" valid:"optional,version"`
```

does not!

**Comment by wangdrew** [comment]
I'm also having the same issue here, would love to know the reasoning!

**Comment by asaskevich** [comment]
Did you attempted reverse order of validators in tag? It seems to be issue when `version` overrides `optional`

**Comment by richardlt** [comment]
Hi, I faced the same issue with custom validators. In fact I just looked at the code in validator.go and I'm not sure to understand why the isEmptyValue check is done after custom validators execution. 

line 727 on master
```golang
	if isEmptyValue(v) {
		// an empty value is not validated, check only required
		return checkRequired(v, t, options)
	}
```

Because in isEmptyValue there is a return at the end that is doing a reflect equal test, I think this should work for all custom types if those lines are moved just after options parsing on line 705.

Thanks for your help.

**Comment by richardlt** [comment]
#185 This is the PR associated to my message

**Comment by andrewwatkin** [comment]
Use case: A custom validator on a struct to check for certain combinations of fields. But any one field can be empty.

The isEmptyValue() short circuit in typeCheck() breaks this.

I don't see why empty fields should be bypassed by custom checks.

**Comment by asaskevich** [comment]
Thank you for your investigation, it seems to be fixed in #329.
Sorry for long awaited response.
